compiler/rustc_codegen_llvm/src/builder.rs

```diff
@@ -231,14 +233,25 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
         let mut bundles = vec![funclet_bundle];
 
         // Set KCFI operand bundle
+        // FIXME(rcvalle): Add support for no_sanitize attribute.
         let is_indirect_call = unsafe { llvm::LLVMIsAFunction(llfn).is_none() };
-        let kcfi_bundle =
-            if self.tcx.sess.is_sanitizer_kcfi_enabled() && let Some(fn_abi) = fn_abi && is_indirect_call {
-                let kcfi_typeid = kcfi_typeid_for_fnabi(self.tcx, fn_abi);
-                Some(llvm::OperandBundleDef::new("kcfi", &[self.const_u32(kcfi_typeid)]))
-            } else {
-                None
-            };
+        let kcfi_bundle = if is_indirect_call
+            && self.tcx.sess.is_sanitizer_kcfi_enabled()
+            && !fn_attrs.no_sanitize.contains(SanitizerSet::KCFI)
+        {
+            let mut options = TypeIdOptions::empty();
+            if self.tcx.sess.is_sanitizer_cfi_generalize_pointers_enabled() {
+                options.insert(TypeIdOptions::GENERALIZE_POINTERS);
+            }
+            if self.tcx.sess.is_sanitizer_cfi_normalize_integers_enabled() {
+                options.insert(TypeIdOptions::NORMALIZE_INTEGERS);
+            }
+
+            let kcfi_typeid = kcfi_typeid_for_fnabi(self.tcx, fn_abi.unwrap(), options);
+            Some(llvm::OperandBundleDef::new("kcfi", &[self.const_u32(kcfi_typeid)]))
```

Comment:
Whether KCFI is enabled and if the GENERALIZE_POINTERS and NORMALIZE_INTEGERS flags are used for a function is part of the ABI, right? Does LLVM allow enabling KCFI for a single function call without LTO or globally enabling KCFI? And is the ABI stable or not? If it is stable and doesn't need to be enabled globally, maybe it would make sense to move it to `FnAbi`? Or just have a flag that enables in in `CodegenFnAttrs` rather than disables it to avoid UB when trying to mix non-KCFI and KCFI compiled crates together. If it needs to be enabled globally, maybe it could be recorded in the crate metadata if KCFI was enabled and the crate loader could be modified to forbid mixing.

Reply:
> Whether KCFI is enabled and if the GENERALIZE_POINTERS and NORMALIZE_INTEGERS flags are used for a function is part of the ABI, right?

I'm not sure if CFI and KCFI are part of any formal ABI, but they're implemented in Clang/LLVM and their documentation are at https://clang.llvm.org/docs/ControlFlowIntegrity.html and https://clang.llvm.org/docs/ControlFlowIntegrityDesign.html.

> Does LLVM allow enabling KCFI for a single function call without LTO or globally enabling KCFI?

It doesn't allow it to be enabled per function (and CFI and KCFI require LTO) but it allows an ignore list (see https://clang.llvm.org/docs/ControlFlowIntegrity.html#ignorelist). This is implemented for Rust with the no_sanitize attribute.

> And is the ABI stable or not?

I'm not sure if they're are part of any formal ABI, but they're implemented in Clang/LLVM and are not experimental features.

> If it is stable and doesn't need to be enabled globally, maybe it would make sense to move it to `FnAbi`?

Enabling it globally is the preferred default (when enabling the feature). The no_sanitize attribute is for disabling CFI and KCFI checks in the caller (and thus applied to the caller) and not the callee, but unfortunately the FnAbi is for the callee in this case.

>  Or just have a flag that enables in in `CodegenFnAttrs` rather than disables it to avoid UB when trying to mix non-KCFI and KCFI compiled crates together. If it needs to be enabled globally, maybe it could be recorded in the crate metadata if KCFI was enabled and the crate loader could be modified to forbid mixing.

Both CFI and KCFI are implemented so that compilation units/crates with CFI or KCFI interoperate/work with compilation units/crates without and their behavior is well documented (see above) so there is no UB in these cases.

Reply:
I don't think including this in FnAbi would work, because what we care about here is the no_sanitize attribute on the caller, not the callee. `fn_abi` is the callee ABI, while `fn_attrs` are the caller attributes (which is not super obvious).

I feel like this is *probably* not the right place to handle this -- can we pull this higher up into the place that has the necessary context -- if I'm seeing that right, it's just do_call that needs this handling, not other places generating calls?

Reply:
I've been pondering about this too. However, do_call is part of rustc_codegen_ssa, which is non LLVM specific, and operand bundles are an LLVM-specific concept (AFAIK?), that is the main reason I decided to implement and left it in rustc_codegen_llvm/src/builder.rs.

I could implement something similar to [Funclet](https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_llvm/src/common.rs#L66) implemented in rustc_codegen_llvm/src/common.rs and used in do_call, but I've the impression that it actually belongs to and at some point should be moved to be used in rustc_codegen_llvm/src/builder.rs as well.

What do you think?

Reply:
I think what we'd want is a trait method like `handle_forward_cfi` that isn't LLVM specific by itself, but for now will only be implemented by codegen_llvm with the necessary kcfi operand bundle handling.

Ideally, that API would also handle the normal CFI type tests, because it's sort of weird that this also very LLVM-specific implementation is part of codegen_ssa. But possibly that's hard to do because it touches control flow. Something I noticed in that context is that the CFI handling is only in codegen_call_terminator, while e.g. codegen_drop_terminator can also perform indirect (vtable) calls.

Reply:
> > It doesn't allow it to be enabled per function (and CFI and KCFI require LTO)
> 
> Just a side note: KCFI does _not_ require LTO, that was one of the motivations for adding KCFI to LLVM. The traditional LLVM CFI does require LTO, of course.

Thanks for pointing this out, Sami! I'll change it accordingly.

Reply:
> I think what we'd want is a trait method like `handle_forward_cfi` that isn't LLVM specific by itself, but for now will only be implemented by codegen_llvm with the necessary kcfi operand bundle handling.

Would this be something different/separate from the abstraction I described below?

To move the KCFI implementation up to codegen_ssa, I'd need to implement something similar to Funclet I mentioned above, since operand bundles are passed as parameters to LLVMRustBuildInvoke in the LLVM Builder call and invoke methods and are not attached to the callee in any way, and change codegen_ssa to pass a vector of wrapped operand bundles to do_call. However, I'm still pondering whether codegen_ssa should be handling operand bundles or if they should be handled by the LLVM Builder since they are LLVM specific (and so is KCFI, which is also different from the traditional CFI). What do you think?

> 
> Ideally, that API would also handle the normal CFI type tests, because it's sort of weird that this also very LLVM-specific implementation is part of codegen_ssa. But possibly that's hard to do because it touches control flow. 

There is the type membership abstraction I added to the backend-agnostic code (i.e., the TypeMembershipMethods supertrait of the TypeMethods subtrait in codegen_ssa). The abstraction was created because current hardware-assisted and software-based -combined forward-edge control flow protections are implemented similarly (i.e., testing whether a given pointer is associated with a type identifier) and as you pointed out because it also touches control flow. Do you think we should change it now to something more abstract or we should wait until another implementation to see if any changes to this abstraction are necessary?

> Something I noticed in that context is that the CFI handling is only in codegen_call_terminator, while e.g. codegen_drop_terminator can also perform indirect (vtable) calls.

Thank for pointing this out! I'll add it there too. Let me know if you come across or see any other areas it'd also be needed. (FYI, I'm planning to do a few source code review and binary analysis passes before submitting the feature for stabilization in the future to close any possible remaining gaps.)

Reply:
> To move the KCFI implementation up to be called from codegen_ssa, I'd need to implement something similar to Funclet I mentioned above, since operand bundles are passed as parameters to LLVMRustBuildInvoke in the LLVM Builder call and invoke methods and are not attached to the callee in any way, and change codegen_ssa to pass a vector of wrapped operand bundles to do_call.

Ah yeah, I totally forgot that the bundles need to be specified in advance. That does make things less simple.

> However, I'm still pondering whether codegen_ssa should be handling operand bundles or if they should be handled by the LLVM Builder since they are LLVM specific (and so is KCFI, which is also different from the traditional CFI). What do you think?

It doesn't necessarily have to use the term "operand bundle" in particular, but it seems like passing in the necessary information in some form (similarly to Funclet) is hard to avoid.

> There is the type membership abstraction I added to the backend-agnostic code (i.e., the TypeMembershipMethods supertrait of the TypeMethods subtrait in codegen_ssa). The abstraction was created because current hardware-assisted and software-based -combined forward-edge control flow protections are implemented similarly by testing type membership (e.g., testing whether a given pointer is associated with a type identifier) and as you pointed out because it also touches control flow. Do you think we should change it now to something more abstract or we should wait until another implementation to see if any changes to this abstraction are necessary?

I don't think any changes here are necessary. My note was just that it would be nice if CFI and KCFI could be handled in the same place (rather than one sitting in codegen_call_terminator and the other in the call builder method).

Reply:
As we discussed, it'd be difficult to move the KCFI implementation up to be called from codegen_ssa, but it turns out that it was relatively straightforward to move the CFI implementation down to rustc_codegen_llvm/src/builder.rs to be in the same location as the KCFI implementation, which looking at it now feels like a more suitable location for implementing it. It not only provides more flexibility to other backends to implement CFI by testing type membership, but also covered codegen_drop_terminator, and even ended optimizing out an additional branch in some cases. Let me know what you think.